resourseleak/TestMessage.java

```diff
@@ -0,0 +1,87 @@
+package resourceleak;
+
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+import java.io.*;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.zip.GZIPOutputStream;
+
+
+public class TestMessage {
+
+    private static Map<String, JsonObject> testMessage1(String directory, String fileNamePrefix) throws IOException {
+        Map<String, JsonObject> modelMeta = new HashMap<>();
+
+        File[] files = new File(directory).listFiles((dir, name) -> name.startsWith(fileNamePrefix));
+        for (File file : files) {
+            try {
+                JsonObject obj = (JsonObject) new JsonParser().parse(new FileReader(file));
+                String modelType = getType(file);
+                modelMeta.put(modelType, obj);
+            } catch (IOException e) {
+                LOGGER.error("Reading model metadata file {} error.", file.toString(), e);
+                throw e;
+            }
+        }
+        return modelMeta;
+    }
+
+    private KeyWriter testMessage2() throws IOException {
+        OutputStream out = new FileOutputStream(outputFile);
+
+        if (outputFile.endsWith(".gz")) {
+            try {
+                out = new GZIPOutputStream(out);
+            } catch (IOException e) {
+                out.close();
+                throw e;
+            }
+        }
+
+        Writer writer = new OutputStreamWriter(out, StandardCharsets.UTF_8);
+
+        KeyWriter keyWriter;
+        if (format.equalsIgnoreCase("sableloadgen")) {
+            keyWriter = new SableLoadGenKeyWriter(writer);
+        } else if (format.equalsIgnoreCase("blackmarlin")) {
+            keyWriter = new BlackMarlinKeyWriter(writer);
+        } else {
+            throw new IllegalArgumentException("format must be either sableloadgen or blackmarlin");
+        }
+
+        return keyWriter;
+    }
+
+    private Writer testMessage3() throws IOException {
+        OutputStream out = new FileOutputStream(outputFile);
+
+        if (outputFile.endsWith(".gz")) {
+            try {
+                out = new GZIPOutputStream(out);
+            } catch (IOException e) {
+                out.close();
+                throw e;
+            }
+        }
+
+        Writer writer = new OutputStreamWriter(out, StandardCharsets.UTF_8);
+
+        if (! format.equalsIgnoreCase("sableloadgen")) {
+            throw new IllegalArgumentException("format must be either sableloadgen or blackmarlin");
+
+        }
+        return writer;
+    }
+
+    private Writer testReturnMinimalFindingsThatSpanAllLines() throws IOException {
+        OutputStream out = new FileOutputStream(outputFile);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *out*. The resource is referenced in statements at the following line: 81. The resource is returned at line: 85. There are other execution paths that don&#39;t close the resource or return it, for example, when an if-branch is taken at line 82. To prevent this resource leak, close **out** along these other paths before you exit this method.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).